juju/juju
# LXDProfile -  Don't apply when not provisioned

## Description of change

Don't apply when not provisioned  …
The following changes ensure that we don't attempt to apply a
charm profile to a machine that isn't either running or that isn't
provisioned.

Interestingly the charm profile will be written later on when ensuring
that the profile is there, by the provisioner API. So we don't have
to write any other code that to stop the applying of the charm
profile.

 - Add tests around provisioner task

## QA steps

Run the following commands to ensure that add machines correctly.
```sh
juju bootstrap lxd
juju deploy ~/bundle-no-sub.yaml
watch --color -n 1 juju status --color
```

You should see the following output in status:
```
Model    Controller  Cloud/Region  Version  SLA          Timestamp
default  xxx         lxd/default   2.5.1.1  unsupported  16:21:56Z

App          Version  Status  Scale  Charm        Store  Rev  OS      Notes
lxd-profile  18.04    active      9  lxd-profile  local    0  ubuntu  
ubuntu       18.04    active      9  ubuntu-lite  local    0  ubuntu  

Unit            Workload  Agent  Machine  Public address  Ports  Message
lxd-profile/0*  active    idle   0        10.178.104.238         load: 0.54, 1.18, 1.39
lxd-profile/1   active    idle   1        10.178.104.176         load: 0.50, 1.16, 1.38
lxd-profile/2   active    idle   2        10.178.104.163         load: 1.13, 1.14, 1.33
lxd-profile/3   active    idle   3        10.178.104.151         load: 1.26, 1.21, 1.35
lxd-profile/4   active    idle   4        10.178.104.226         load: 1.28, 1.17, 1.35
lxd-profile/5   active    idle   5        10.178.104.253         load: 1.26, 1.21, 1.35
lxd-profile/6   active    idle   6        10.178.104.47          load: 1.28, 1.17, 1.35
lxd-profile/7   active    idle   7        10.178.104.43          load: 1.16, 1.19, 1.34
lxd-profile/8   active    idle   8        10.178.104.185         load: 1.28, 1.17, 1.35
ubuntu/0*       active    idle   0        10.178.104.238         load: 1.06, 1.16, 1.33
ubuntu/1        active    idle   1        10.178.104.176         load: 1.13, 1.14, 1.33
ubuntu/2        active    idle   2        10.178.104.163         load: 0.66, 1.15, 1.37
ubuntu/3        active    idle   3        10.178.104.151         load: 0.66, 1.15, 1.37
ubuntu/4        active    idle   4        10.178.104.226         load: 1.13, 1.18, 1.33
ubuntu/5        active    idle   5        10.178.104.253         load: 0.66, 1.15, 1.37
ubuntu/6        active    idle   6        10.178.104.47          load: 0.61, 1.13, 1.37
ubuntu/7        active    idle   7        10.178.104.43          load: 1.13, 1.18, 1.33
ubuntu/8        active    idle   8        10.178.104.185         load: 1.13, 1.18, 1.33

Machine  State    DNS             Inst id        Series  AZ  Message
0        started  10.178.104.238  juju-c02210-0  bionic      Running
1        started  10.178.104.176  juju-c02210-1  bionic      Running
2        started  10.178.104.163  juju-c02210-2  bionic      Running
3        started  10.178.104.151  juju-c02210-3  bionic      Running
4        started  10.178.104.226  juju-c02210-4  bionic      Running
5        started  10.178.104.253  juju-c02210-5  bionic      Running
6        started  10.178.104.47   juju-c02210-6  bionic      Running
7        started  10.178.104.43   juju-c02210-7  bionic      Running
8        started  10.178.104.185  juju-c02210-8  bionic      Running
```

And the `lxc profile list` should also state:
```
+----------------------------+---------+
|            NAME            | USED BY |
+----------------------------+---------+
| default                    | 10      |
+----------------------------+---------+
| juju-controller            | 1       |
+----------------------------+---------+
| juju-default               | 9       |
+----------------------------+---------+
| juju-default-lxd-profile-0 | 8       |
+----------------------------+---------+
``` 

bundle-no-sub.yaml
```yaml
series: bionic
applications:
  lxd-profile:
    charm: ~/go/src/github.com/juju/juju/testcharms/charm-repo/quantal/lxd-profile
    num_units: 9
    to:
    - "0"
    - "1"
    - "2"
    - "3"
    - "4"
    - "5"
    - "6"
    - "7"
    - "8"
  ubuntu:
    charm: cs:~jameinel/ubuntu-lite
    num_units: 9
    to:
    - "0"
    - "1"
    - "2"
    - "3"
    - "4"
    - "5"
    - "6"
    - "7"
    - "8"
machines:
  "0": {}
  "1": {}
  "2": {}
  "3": {}
  "4": {}
  "5": {}
  "6": {}
  "7": {}
  "8": {}
```

## Documentation changes



## Bug reference
https://bugs.launchpad.net/juju/+bug/1813044
